### PR TITLE
Timediff: Duration between two datetimes in a variety of units

### PR DESCRIPTION
**Pull request by mintsoft**
I found myself looking through a variety of logs today and there's no IA for finding how long it is between two points in time, we have one for days etc but not timestamps accurate to the second. Hence this IA

timediff

-----------

https://duck.co/ia/view/timediff

### CODE REVIEW

**Comment by daxtheduck** [comment]
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

**Comment by GuiltyDolphin** [comment]
@mintsoft Sweet, yeah - I was just thinking the other day "I want *minutes* between 2am and 5pm". Presumably this would deprecate `DaysBetween`?

File: lib/DDG/Goodie/Timediff.pm [hunk 1]
```diff
@@ -0,0 +1,49 @@
+package DDG::Goodie::Timediff;
+# ABSTRACT: provides the duration between dates
+
+use strict;
+use DDG::Goodie;
+with 'DDG::GoodieRole::Dates';
+with 'DDG::GoodieRole::NumberStyler';
+
+triggers any => qw(timediff);
```
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/Timediff.pm]: Triggers like `time difference`, `minutes/seconds/hours (etc) between`, `time between` would all be useful.

File: lib/DDG/Goodie/Timediff.pm [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+zci is_cached   => 0;
+zci answer_type => 'timediff';
+
+my $datestring_regex = datestring_regex();
+my $style = number_style_for("1");
```
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/Timediff.pm]: Cheeky! Looks like you need [something that would do number formatting based on locale](https://github.com/duckduckgo/zeroclickinfo-goodies/blob/gd/numberstyler-improvements/lib/DDG/GoodieRole/NumberStyler.pm#L26)!

**Comment by mintsoft** [comment]
@GuiltyDolphin can you cast your eye over the `structured_answer` for me? I'm getting:
`Use of uninitialized value in concatenation (.) or string at /home/rob/perl5/lib/perl5/App/DuckPAN/Web.pm line 447.` in `duckpan server` but I just can't see what I've done wrong

File: lib/DDG/Goodie/Timediff.pm (hunk 2 above)
Comment (mintsoft) [inline lib/DDG/Goodie/Timediff.pm]: Indeed! That's exactly what I'm after

**Comment by mintsoft** [comment]
@GuiltyDolphin Nevermind, it's gone away

File: lib/DDG/Goodie/Timediff.pm [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    
+    my @dates = parse_all_datestrings_to_date($1,$2);
+    return unless((scalar @dates) == 2);
+    my $duration = abs $dates[0]->epoch - $dates[1]->epoch;
+    
+    return "$duration seconds",
+        structured_answer => {
+            data => {
+                record_data => {
+                    seconds => $style->for_display($duration),
```
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/Timediff.pm]: For me eyes - can we line these up? I'm sure Atom (you use Atom?) has *something* like Tabularize...

File: lib/DDG/Goodie/Timediff.pm [hunk 4]
```diff
@@ -0,0 +1,47 @@
+package DDG::Goodie::Timediff;
+# ABSTRACT: provides the duration between dates
+
+use strict;
+use DDG::Goodie;
+with 'DDG::GoodieRole::Dates';
+with 'DDG::GoodieRole::NumberStyler';
+
+triggers any => ["timediff", "time difference", "minutes between", "seconds between","hours between", "duration between"];
```
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/Timediff.pm]: How about doing something like...

```perl
# We could use these later on as well, possibly to change which results are displayed?
my @units = (
    "seconds",
    "minutes",
    "hours",
    "days",
    ...
);
triggers any => (..., map { "$_ between" } @units);
```

File: lib/DDG/Goodie/Timediff.pm [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    return "$duration seconds",
+        structured_answer => {
+            data => {
+                record_data => {
+                    seconds => $style->for_display($duration),
+                    minutes => $style->for_display($duration/60),
+                    hours => $style->for_display($duration/3600),
+                    days => $style->for_display($duration/86400)
+                },
+                record_keys => ['days', 'hours', 'minutes', 'seconds'],
```
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/Timediff.pm]: Mentioned in above comment, we could have `@units` for these.

File: t/Timediff.t [hunk 6]
```diff
@@ -0,0 +1,58 @@
+#!/usr/bin/env perl
+use strict;
+use warnings;
+use Test::More;
+use DDG::Test::Goodie;
+use utf8;
+zci answer_type => 'timediff';
+zci is_cached   => 0;
+
+sub make_answer(%){
```
Comment (GuiltyDolphin) [inline t/Timediff.t]: Woah, some protos going on here, eh?

File: t/Timediff.t [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+    'timediff 2016-04-11T09:00:00 2016-04-08T18:17:00' => test_zci(
+        "225780 seconds",
+        structured_answer => make_answer({
+			seconds => "225,780",
+			minutes => "3,763",
+			hours => "62.7166666666667",
+			days => "2.61319444444444",
+        })
+    ),
+    'timediff 2016-04-08T20:00:00 2016-04-10T21:00:00' => test_zci(
```
Comment (GuiltyDolphin) [inline t/Timediff.t]: Maw tests please. Lets have some for descriptive dates + other triggers.

File: lib/DDG/Goodie/Timediff.pm [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+    my $duration = abs $dates[0]->epoch - $dates[1]->epoch;
+    
+    return "$duration seconds",
+        structured_answer => {
+            data => {
+                record_data => {
+                    seconds => $style->for_display($duration),
+                    minutes => $style->for_display($duration/60),
+                    hours => $style->for_display($duration/3600),
+                    days => $style->for_display($duration/86400)
```
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/Timediff.pm]: I was thinking, if the user does something like `seconds between 1st Jan 2012 and 2nd Jan 2012 00:07:31`, should we only display seconds? Obviously displaying the other units doesn't hurt but... If someone is specific enough it feels like displaying only the specified unit would make more sense.

File: lib/DDG/Goodie/Timediff.pm (hunk 3 above)
Comment (mintsoft) [inline lib/DDG/Goodie/Timediff.pm]: I actually use VS Code these days!

File: lib/DDG/Goodie/Timediff.pm (hunk 4 above)
Comment (mintsoft) [inline lib/DDG/Goodie/Timediff.pm]: Urr I can do, but it'll be way more code than is already there!

Comment (GuiltyDolphin) [inline lib/DDG/Goodie/Timediff.pm]: @mintsoft I've seen the units used in a few places, we can just replace occurrences with the list (may have to re-order the list to your liking).

File: t/Timediff.t (hunk 7 above)
Comment (mintsoft) [inline t/Timediff.t]: I'll focus more on the triggers I think; testing the date parsing responsibility in detail isn't this IA's responsibility. It falls to the Role's Tests, however I agree we should have some tests guaranteeing the expected contract between the role and this IA

File: lib/DDG/Goodie/Timediff.pm (hunk 8 above)
Comment (mintsoft) [inline lib/DDG/Goodie/Timediff.pm]: For me, the whole idea was to get a feeling of how much of each unit is. I'm getting really annoyed by the recent movement towards "just now" and "4 hours ago" as durations as if that's accurate enough for someone to be able actually determine when events took place.

**Comment by mintsoft** [comment]
@GuiltyDolphin WTF are the tests broken in master? https://travis-ci.org/duckduckgo/zeroclickinfo-goodies/builds/121914770#L417

File: t/Timediff.t (hunk 7 above)
Comment (GuiltyDolphin) [inline t/Timediff.t]: I mean like, `how many minutes between 3rd Jan 2012 and 5th Jan 2015`, i.e, things people might type. Of course not isolating these tests - no point in *just* testing date format.

File: t/Timediff.t [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+        "3600 seconds",
+        structured_answer => make_answer({
+            seconds => "3,600",
+            minutes => "60",
+            hours => "1",
+            days => "0.0416666666666667",
+        })
+    ),
+    
+    'days between 13/05/2016 and 01/06/2016' => test_zci(
```
Comment (GuiltyDolphin) [inline t/Timediff.t]: Jus' so you know - we'll need to wrap this in a Language/Location test when #2810 gets merged - I think default ('Murican) format won't accept this.

File: t/Timediff.t [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+        "1641600 seconds",
+        structured_answer => make_answer({
+            seconds => "1,641,600",
+            minutes => "27,360",
+            hours => "456",
+            days => "19",
+        })
+    ),
```
Comment (GuiltyDolphin) [inline t/Timediff.t]: How about a couple of time-dependent tests? For example, `time between now and tomorrow`, `seconds between now and Jan 1st`.

File: t/Timediff.t (hunk 9 above)
Comment (mintsoft) [inline t/Timediff.t]: Yeah that's cool; it's basically why I put this one in :)

File: t/Timediff.t (hunk 10 above)
Comment (mintsoft) [inline t/Timediff.t]: I've just added some relative tests; however there is a bug with the dates role. If "today","yesterday","tomorrow" is used, then it doesn't round off the time; so I've not included tests that exhibit the Role's bug. Do you know if you happened to fix this in the refactor?

Comment (GuiltyDolphin) [inline t/Timediff.t]: "Refactor", I've changed a lot of the parsing functionality, but I think tomorrow, yesterday etc. will be tomorrow or yesterday from the current ***time***.

Comment (mintsoft) [inline t/Timediff.t]: I beg your pardon; "changes" then. I think `tomorrow` and `yesterday` should be relative to `today` rather than `now`

Comment (GuiltyDolphin) [inline t/Timediff.t]: @mintsoft Yeah, makes sense. I guess if you want the time tomorrow you can just do `time tomorrow` or in the Goodie have `now` then + 24 hours.

**Comment by mintsoft** [comment]
@moollaza @zachthompson @GuiltyDolphin  any idea why this hasn't deployed?

**Comment by GuiltyDolphin** [comment]
@mintsoft Deploys are manual for full IAs AFAIK; @moollaza this should be ready for Beta.

**Comment by mintsoft** [comment]
@GuiltyDolphin Ah fair enough; I thought that'd been automated. I've obviously been looking at too many cheatsheets!

File: t/Timediff.t [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+			hours => "62.7166666666667",
+			days => "2.61319444444444",
+            in1 => "11 Apr 2016 09:00:00 EDT",
+            in2 => "08 Apr 2016 18:17:00 EDT"
+        })
+    ),
+    'timediff 2016-04-08T20:00:00 2016-04-10T21:00:00' => test_zci(
+        "176400 seconds",
+        structured_answer => make_answer({
+			seconds => "176,400",
```
Comment (GuiltyDolphin) [inline t/Timediff.t]: NGHHH TABS???

Comment (mintsoft) [inline t/Timediff.t]: Whoops; I must have accidentically switched my editor over into sensible mode. I'll put it back.

**Comment by moollaza** [comment]
@mintsoft apologies for the delay! This LGTM :+1:

File: t/Timediff.t [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+
+ddg_goodie_test(
+    ['DDG::Goodie::Timediff'],
+
+    'timediff 2016-04-11T09:00:00 2016-04-08T18:17:00' => test_zci(
+        "225780 seconds",
+        structured_answer => make_answer({
+            seconds => "225,780",
+            minutes => "3,763",
+            hours => "62.7166666666667",
```
Comment (moollaza) [inline t/Timediff.t]: maybe we should round these?

**Comment by daxtheduck** [comment]
@mintsoft this is now merged! It should be live on DuckDuckGo within one week (or during the next release) and we'll be sure to notify you when it happens.

After your Instant Answer goes live for the World to use, you will be able to see [traffic information](http://docs.duckduckhack.com/maintaining/tools.html#monitor-your-instant-answer-page) about your IA as well as receive feedback directly from users. 
                
Please refer to the [maintenance guide and docs](http://docs.duckduckhack.com/maintaining/guidelines.html) to keep up with potential changes to your Instant Answer.
                
Thank you again for your help!

File: lib/DDG/Goodie/Timediff.pm [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+handle remainder => sub {
+    my $query = $_;
+
+    return unless $query =~ /^(?:between\s)?($datestring_regex)\s(?:and\s)?($datestring_regex)$/i;
+    
+    my @dates = parse_all_datestrings_to_date($1,$2);
+    return unless((scalar @dates) == 2);
+    my $duration = abs $dates[0]->epoch - $dates[1]->epoch;
+    
+    return "$duration seconds",
```
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/Timediff.pm]: @mintsoft Hmm... Maybe we could return a formatted string as the text response? E.g., `5 days, 3 hours, 7 minutes, and 8 seconds`? `DateTime::Format::Human::Duration` should be able to handle this.

Comment (mintsoft) [inline lib/DDG/Goodie/Timediff.pm]: I wanted it as a fraction deliberately :)

Comment (GuiltyDolphin) [inline lib/DDG/Goodie/Timediff.pm]: @mintsoft You mean as an integer?

Comment (mintsoft) [inline lib/DDG/Goodie/Timediff.pm]: Yeah integer sorry

Comment (GuiltyDolphin) [inline lib/DDG/Goodie/Timediff.pm]: @mintsoft Fair do's - it's a fairly sensible format.

**Pull request merged**